lm_human_preference_details/train_policy_jax.py

```diff
@@ -0,0 +1,1024 @@
+                + f"[red]score: {sample_score}, kl: {sample_kl}, total reward: {sample_score - kl_ctl.value * sample_kl} [/]"
+            )
+        except Exception as e:
+            print(e)
+
+        # Rollout metrics
+        rl_stats = rl_stats.unreplicate().compute()
+        non_score_reward = -kl_ctl.value * kl
+
+        mean_kl = kl.sum(-1).mean()
+        mean_entropy = (-logprobs).sum(-1).mean()
+        mean_non_score_reward = non_score_reward.sum(-1).mean()
+        writer.add_scalar("objective/kl_coef", np.array(kl_ctl.value), update)
+        writer.add_scalar("objective/kl", mean_kl.item(), update)
+        writer.add_scalar("objective/entropy", mean_entropy.item(), update)
+        writer.add_scalar(
+            "objective/non_score_reward", mean_non_score_reward.item(), update
+        )
+        writer.add_scalar(
+            "objective/score_total",
+            mean_non_score_reward.item() + scores.mean().item(),
+            update,
+        )
+        writer.add_scalar("objective/scores", scores.mean().item(), update)
+        writer.add_scalar("ppo/returns/mean", returns.mean().item(), update)
+        writer.add_scalar("ppo/returns/var", returns.var().item(), update)
+        writer.add_scalar("ppo/val/mean", values.mean().item(), update)
+        writer.add_scalar("ppo/val/var", values.var().item(), update)
+        writer.add_scalar("ppo/val/advantage", advantages.mean().item(), update)
+        writer.add_scalar(
+            "ppo/val/num_eos_tokens",
+            (query_responses[:, :, args.task.query_length :] == tokenizer.eos_token_id).sum().item(),
+            update,
+        )
+
+        # RL metrics aggregated at the batch level
+        writer.add_scalar(
+            "ppo/policy/approxkl_avg", rl_stats["approxkl"].item(), update
+        )
+        writer.add_scalar(
+            "ppo/policy/entropy_avg",
+            rl_stats["entropy"].item(),
+            update,
+        )
+        writer.add_scalar(
+            "ppo/loss/policy_avg",
+            rl_stats["pg_loss"].item(),
+            update,
+        )
+        writer.add_scalar(
+            "ppo/policy/clipfrac_avg",
+            rl_stats["pg_clipfrac"].item(),
+            update,
+        )
+        writer.add_scalar(
+            "ppo/val/error",
+            rl_stats["vf_loss1"].item(),
+            update,
+        )
+        writer.add_scalar(
+            "ppo/loss/value_avg",
+            rl_stats["vf_loss"].item(),
+            update,
+        )
+        writer.add_scalar(
+            "ppo/val/clipfrac_avg",
+            rl_stats["vf_clipfrac"].item(),
+            update,
+        )
+        writer.add_scalar(
+            "ppo/val/ratio_avg",
+            rl_stats["ratio"].item(),
+            update,
+        )
+        writer.add_scalar("ppo/loss/total", rl_stats["loss"].item(), update)
+
+        # Logging learning rate and learning progress
+        lrnow = linear_schedule(policy_state.step - 1, args)
```

Comment:
learning rate annealing happening outside of `otpax`is a big no no. I remember seeing it makes the training twice as slow. Instead do https://github.com/vwxyzjn/cleanba/blob/81dca0054c8c0930046a2d12b07ada2945014d01/cleanba/cleanba_ppo.py#L497

Reply:
I see, thanks! But here we just call `linear_schedule` to record the learning rate for the metric writer; `lrnow` is only used in metric recording and not used for annealing. But indeed since `policy_state.step` is a jax array, computing it outside of jit can be slow. I replace it by `global_step`, which is just a float.

Reply:
Thanks! Here I called `linear_schedule` to compute the current lr simply for logging purposes. The annealing process is managed by `optax.MultiSteps`, and the computed learning rate here isn't used in annealing. So I'm not sure if it's a significant bottleneck. But I find the solution you recommended to be very elegant---in this way we can directly extract the lr, without recomputing it! I've implemented the change.